src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirRequestContextDocumentClientExtensions.cs

```diff
@@ -0,0 +1,77 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Globalization;
+using System.Net;
+using EnsureThat;
+using Microsoft.Azure.Documents;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using Microsoft.Health.Fhir.Core.Features.Context;
+
+namespace Microsoft.Health.Fhir.CosmosDb.Features.Storage
+{
+    internal static class FhirRequestContextDocumentClientExtensions
+    {
+        /// <summary>
+        /// Adds request charge to the response headers and throws a <see cref="RequestRateExceededException"/>
+        /// if the status code is 429.
+        /// </summary>
+        /// <param name="requestContext">The request context</param>
+        /// <param name="ex">The exception</param>
+        public static void ProcessException(this IFhirRequestContext requestContext, Exception ex)
+        {
+            if (requestContext == null)
+            {
+                return;
+            }
+
+            EnsureArg.IsNotNull(ex, nameof(ex));
+
+            if (ex is DocumentClientException dce)
+            {
+                requestContext.AddRequestChargeToResponseHeaders(dce.RequestCharge);
+
+                if (dce.StatusCode == (HttpStatusCode)429)
+                {
+                    throw new RequestRateExceededException(dce.RetryAfter);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Updates the response headers with the session token and request change values.
+        /// </summary>
+        /// <param name="requestContext">The request context</param>
+        /// <param name="sessionToken">THe session token</param>
+        /// <param name="responseRequestCharge">The request charge.</param>
+        public static void UpdateResponseHeaders(this IFhirRequestContext requestContext, string sessionToken, double responseRequestCharge)
+        {
+            if (requestContext == null)
+            {
+                return;
+            }
+
+            if (!string.IsNullOrEmpty(sessionToken))
+            {
+                requestContext.ResponseHeaders[CosmosDbHeaders.SessionToken] = sessionToken;
+            }
+
+            requestContext.AddRequestChargeToResponseHeaders(responseRequestCharge);
+        }
+
+        private static void AddRequestChargeToResponseHeaders(this IFhirRequestContext requestContext, double responseRequestCharge)
+        {
+            if (requestContext.ResponseHeaders.TryGetValue(CosmosDbHeaders.RequestCharge, out StringValues existingValues) &&
```

Comment:
Instead of storing the value in the header and parsing header back to double, we could store the double value in requestContext itself and then when the value is updated just calculates the new double value and at the same time write to the header?

Reply:
Added a comment explaining the rationale.